### DISCUSSION TITLE
Auto-materialize is skipped and does not trigger parent materialization

### DISCUSSION THREAD

**Discussion by doliveira-lts** [dagster-io/dagster#17963] [Q&A]
Hi all,

I have 3 assets configured on Dagster Cloud in the following matter:

A -> B -> C

Where A and B are multi-partitioned assets (same partition schema) and C is non-partitioned. I have a job that triggers both A and B assets at 8:30am, every day, and C has an Auto-Materialize policy attached (eager).

When any partitions from A/B fails, C is never triggered. Am I missing something here? From the auto-materialize history tab of asset C, I can see that the run was skipped at 8:32am (due to "_Waiting on upstream data to be up to date_", but the job actually started executing only at 8:36am (due "_Still waiting for compute resources to spin up._").

What Am I missing here? Why did the auto-materialize policy was triggered before any asset materialization? Why doesn't it get triggered if any partition fails, since it technically fits all auto-materialize rules?

Thanks!

**Comment by doliveira-lts** [comment]
Maybe a simpler way of framing this problem is:

When a partitioned asset is run and only 90% of partitions succeed, I still want to run downstream assets. How can I configure this with Auto-materialization policies?

**Reply by sryza** [reply]
Ah, this explains what's going on. The default policy expect that all upstream partitions have a partition before materializing the downstream asset.

What behavior would you ideally want? For the downstream asset to materialize as soon as 1 of the partitions of the upstream asset is materialized? As soon as none of the upstream partitions are in progress?

**Reply by doliveira-lts** [reply]
I would like the downstream asset to materialize as soon as none of the upstream partitions are in progress (eg the "run" has finished), either by succeeding or failing. I might also have missing partitions from previous runs (I'm using the yield Output in some cases).

Ideally, a % threshold of failed tasks would also be great (but this is easy to test on the downstream asset).

**Reply by sryza** [reply]
Got it - this isn't possible at the moment, but we're currently designing an API that will allow specifying custom auto-materialize conditions.

One followup on "I would like the downstream asset to materialize as soon as none of the upstream partitions are in progress": does this apply only to direct parents? Or would you want the downstream asset not to be materialized even if ancestors were in progress?

**Reply by doliveira-lts** [reply]
I think this would include ancestors as well, since in production I actually have 2 upstream assets with the same partitions. Ideally the downstream asset would only run when all partitions from all ancestors are done, either succeeding or failing. 

As a temporary fix, I hardcoded a second schedule to run the downstream asset some time after all partitions would finish, but it's not ideal.

**Comment by doliveira-lts** [comment]
Here's a reproductible example of the problem (dagster 1.5.7):

```python
@asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-11-10")
)
def asset_A(context):
    partition_date_str = context.asset_partition_key_for_output()
    if partition_date_str == '2023-11-10':
        print('Waiting....')
        time.sleep(45)
    if partition_date_str == '2023-11-11':
        time.sleep(120)
        print('Breaking....')
        raise
   

@asset(
    auto_materialize_policy=AutoMaterializePolicy.eager()
)
def asset_B(context, asset_A):
    pass
```

When running all partitions of `asset_A` with the 'Materialize All' feature, `asset_B` will not be triggered.

**Reply by sryza** [reply]
Hey @doliveira-lts - if you go to the Auto-materialize history tab of the page for `asset_B`, does it give you a reason for why it's skipped?

**Reply by doliveira-lts** [reply]
I can see that the run was skipped through the webserver log on my terminal, but nothing is shown on the UI ("_No conditions met_")

**Reply by sryza** [reply]
Sorry for the slow response

> I can see that the run was skipped through the webserver log on my terminal

Are you able to share the output?

> but nothing is shown on the UI ("No conditions met")

If you look earlier in history, do you see evidence of skips?
